metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/WorkflowExecutionMonitor.java

```diff
@@ -0,0 +1,335 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.redisson.api.RLock;
+import org.redisson.api.RedissonClient;
+import org.redisson.client.RedisConnectionException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import eu.europeana.metis.core.dao.WorkflowExecutionDao;
+import eu.europeana.metis.core.rest.ResponseListWrapper;
+import eu.europeana.metis.core.workflow.OrderField;
+import eu.europeana.metis.core.workflow.WorkflowExecution;
+import eu.europeana.metis.core.workflow.WorkflowStatus;
+
+/**
+ * This class monitors workflow executions. It provides functionality that determines whether a
+ * running execution is progressing (as opposed to hanging) as well as functionality to deal with
+ * this eventuality. Two of its methods are meant to be scheduled for a periodical run.
+ */
+public class WorkflowExecutionMonitor {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(WorkflowExecutionMonitor.class);
+
+  private static final String FAILSAFE_LOCK = "failsafeLock";
+
+  private final WorkflowExecutionDao workflowExecutionDao;
+  private final WorkflowExecutorManager workflowExecutorManager;
+  private final Duration failsafeLeniency;
+  private final RLock lock;
+
+  /** The currently running executions. **/
+  private Map<String, WorkflowExecutionEntry> currentExecutions = Collections.emptyMap();
+
+  /**
+   * Constructor the executor
+   *
+   * @param workflowExecutorManager {@link WorkflowExecutorManager}
+   * @param workflowExecutionDao {@link WorkflowExecutionDao}
+   * @param redissonClient {@link RedissonClient}
+   * @param failsafeLeniency The leniency given to executions to be idle.
+   */
+  public WorkflowExecutionMonitor(WorkflowExecutorManager workflowExecutorManager,
+      WorkflowExecutionDao workflowExecutionDao, RedissonClient redissonClient,
+      Duration failsafeLeniency) {
+    this.failsafeLeniency = failsafeLeniency;
+    this.workflowExecutionDao = workflowExecutionDao;
+    this.workflowExecutorManager = workflowExecutorManager;
+    this.lock = redissonClient.getFairLock(FAILSAFE_LOCK);
+  }
+
+  /**
+   * Makes a run to make an inventory of the last update times of all running tasks. This method is
+   * meant to run periodically.
+   */
+  public void performExecutionMonitoring() {
+    try {
+
+      // Lock for the duration of this scheduled task.
+      lock.lock();
+
+      // Update the current executions map (completely replace it).
+      updateCurrentExecutions();
+
+    } catch (RuntimeException e) {
+      LOGGER.warn("Exception thrown from Redis, monitoring thread continues", e);
+    } finally {
+      try {
+        lock.unlock();
+      } catch (RedisConnectionException e) {
+        LOGGER.warn("Cannot connect to unlock, monitoring thread continues", e);
+      }
+    }
+  }
+
+  List<WorkflowExecution> updateCurrentExecutions() {
+
+    // Check lock
+    if (!lock.isHeldByCurrentThread()) {
```

Comment:
Are all these checks for current thread having the lock required?

Reply:
Maybe not, but I wanted to be extra sure, also in the case of further changes in this class. Also, since these methods are not private (because I wanted to unit-test them) I wanted to be protective.

Reply:
I have looked into this and it turns out that this method actually talks to the server. So for now I have removed these checks.